### PR TITLE
[CARBONDATA-2804][DataMap] fix the bug when bloom filter or preaggregate datamap tried to be created on older V1-V2 version stores

### PR DESCRIPTION
**Pull request by ndwangsen**
[CARBONDATA-2804] fix the bug when bloom filter or preaggregate datamap tried to be created on older V1-V2 version store

This PR change read carbon file version from carbondata file header to carbonindex file header, because
the version field of carbondata file header is not compatible with  older v1/v2 version. 

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 NA
 - [ ] Any backward compatibility impacted?
 NA
 - [ ] Document update required?
NA
 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       test pass in test environment
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
NA

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7731/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6457/

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java [hunk 1]
```diff
@@ -3212,28 +3213,27 @@ public static ColumnarFormatVersion getFormatVersion(CarbonTable carbonTable)
...
-            }
-            return file.getName().endsWith("carbondata");
-          }
-        });
-    if (carbonFiles == null || carbonFiles.length < 1) {
-      return CarbonProperties.getInstance().getFormatVersion();
+    // get the carbon index file header
+    FileFactory.FileType fileType = FileFactory.getFileType(storePath);
+    ColumnarFormatVersion version = null;
+    if (FileFactory.isFileExist(storePath, fileType)) {
```
Comment (manishgupta88) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: Rename storePath to segmentPath

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7737/

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java [hunk 2]
```diff
@@ -3212,28 +3213,27 @@ public static ColumnarFormatVersion getFormatVersion(CarbonTable carbonTable)
...
+      Map<String, byte[]> carbonIndexMap = fileStore.getCarbonIndexMap();
+      if (carbonIndexMap.size() == 0) {
+        version = CarbonProperties.getInstance().getFormatVersion();
+      }
+      CarbonIndexFileReader indexReader = new CarbonIndexFileReader();
+      for (byte[] fileData : carbonIndexMap.values()) {
+        indexReader.openThriftReader(fileData);
+        IndexHeader indexHeader = indexReader.readIndexHeader();
+        version = ColumnarFormatVersion.valueOf((short)indexHeader.getVersion());
+        break;
```
Comment (manishgupta88) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: once reading is complete close indexReader using try and finally block
indexReader.closeThriftReader()

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java [hunk 3]
```diff
@@ -3212,28 +3213,27 @@ public static ColumnarFormatVersion getFormatVersion(CarbonTable carbonTable)
...
+      }
+    } else {
+      version = CarbonProperties.getInstance().getFormatVersion();
```
Comment (manishgupta88) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: 1. This else condition is not required as for a valid segment its path will exist in the file system.
2. If at all the path does not exist then read the version info from the next valid segment and log a warning that valid segment path does not exist in the system

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6463/

**Comment by xuchuanyin** [comment]
hi, @ndwangsen Can you provide simple description about your solution in the PR description.

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6122/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6127/

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java (hunk 1 above)
Comment (ndwangsen) [reply core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: modified according to review comments

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java (hunk 2 above)
Comment (ndwangsen) [reply core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: modified according to review comments

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java (hunk 3 above)
Comment (ndwangsen) [reply core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: modified according to review comments

**Comment by ndwangsen** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6477/

**Comment by ndwangsen** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7753/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6478/

**Comment by ndwangsen** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7758/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6483/

**Comment by manishgupta88** [comment]
LGTM

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6139/

**Pull request closed without merging**